### ISSUE TITLE
Create clusters in data to group points by similarity

### ISSUE THREAD

**Issue by raphaeltimbo** [petrobras/ccp#29]
In the evaluation module, the user needs to pass a dataframe with historical data for performance calculation.
We need to convert the design curves to the operating condition, but since this process is this process is computationally intensive, it would be nice to have a reduced number of conversions.
One way to achieve this would be to group operational points in clusters based on similarity (vs/vd and mach?) and convert the curves for the mean value of each cluster.

**Comment by raphaeltimbo** [comment]
The calculated converted speed is dependent on the mach number, and therefore on the speed of sound for the suction condition, as shown by the code below:

```python
df = session.df.copy()
df = df[:1700]
df = df.rename(
    columns={
        "P77_PIT-1254034A": "ps",
        "P77_TIT-1254011A": "Ts",
        "P77_PIT-1254032A": "pd",
        "P77_TIT-1254009A": "Td",
        "P77_FIT-1254009A": "delta_p",
        "P77_KT-1254501A": "speed",
    }
)

df = df[df['speed'] > 10]

# Calculate mass flow
df["flow_m"] = 0
df["vs"] = 0
df["speed_sound"] = 0
df["speed_converted"] = 0
points_converted = []

fluid_isis = {
    "methane": 44.04,
    "ethane": 3.18,
    "propane": 0.66,
    "n-butane": 0.15,
    "i-butane": 0.05,
    "n-pentane": 0.03,
    "i-pentane": 0.02,
    "n2": 0.25,
    "h2s": 0.06,
    "co2": 51.55,
}

D = Q_(0.590550, "m")
d = Q_(0.366130, "m")

for i, row in df.iterrows():
    p1 = Q_(row.ps, "bar")
    T1 = Q_(row.Ts, "degC")
    delta_p = Q_(row.delta_p, "mmh2o")
    state = ccp.State(p=p1, T=T1, fluid=fluid_isis)
    fo = ccp.FlowOrifice(state, delta_p, D, d, tappings="flange")
    p_new = ccp.Point.convert_from(p0, suc=state)
    points_converted.append(p_new)
    df.loc[i, "flow_m"] = fo.qm.m
    df.loc[i, "vs"] = state.v().m
    df.loc[i, "speed_sound"] = state.speed_sound().m   
    df.loc[i, "speed_converted"] = p_new.speed.m

fig = px.scatter(df, x="speed_sound", y="speed_converted", trendline="ols")
fig.show()


```

![image](https://github.com/petrobras/ccp/assets/18506378/2f5f79ac-50af-4e48-a5fa-52d85082ac9f)

Therefore, we can separate our clusters based on the speed of sound calculated for each suction condition.

**Comment by raphaeltimbo** [comment]
Code to create each cluster:

```python
from sklearn.cluster import KMeans
# Convert DataFrame to matrix
mat = df['speed_sound'].values.reshape(-1, 1)
# Using sklearn
km = KMeans(n_clusters=5)
km.fit(mat)
# Get cluster assignment labels
labels = km.labels_
# Format results as a DataFrame
# results = pd.DataFrame(data=labels, columns=['cluster'], index=df.index)
df['cluster'] = labels
```

Next step is to create a converted impeller to each cluster and when looping through the data to calculate delta_eff, select the respective impeller for comparison.

**Comment by raphaeltimbo** [comment]
Closed with dd4af44.
